### PR TITLE
Add utility function to save context

### PR DESCRIPTION
**Pull request by nsavoire**
# What does this PR do?

Add utility function to save context (registers + stack) (registers + stack) necessary to perform remote unwinding.
Only works for x86_64 for now, I will do the implementation for ARM later.

Getting a green build was a battle with cppcheck / address sanitizer:
https://gitlab.ddbuild.io/DataDog/ddprof/-/pipelines/7440633

I added a third-party implementation for std::span because gcc 9 does not provide it. We could switch to gcc 11 on 16.04 but this would require compiling our own gcc.

Also added support for google-benchmark.

### CODE REVIEW

File: LICENSE-3rdparty.csv [hunk 1]
```diff
@@ -3,3 +3,4 @@ elfutils,"https://sourceware.org/elfutils/",LGPL-3.0-or-later,Free Software Foun
 libcap,"https://git.kernel.org/pub/scm/libs/libcap/libcap.git/",BSD-3-Clause,Free Software Foundation 
 liblzma,"https://github.com/kobolabs/liblzma",Public-Domain,NA
 llvm-demangle,"https://github.com/llvm/llvm-project",Apache-2.0,llvm-project
+tcbrindle/span,"https://github.com/tcbrindle/span",Boost Software License 1.0
```
Comment (r1viollet) [inline LICENSE-3rdparty.csv:6]: Ah yes, gcc 9 is too old so we pull in tcbrindle. 👍

File: test/savecontext-ut.cc [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+void funcA() __attribute__((noinline));
+void funcB() __attribute__((noinline));
+
+std::byte stack[PERF_SAMPLE_STACK_SIZE];
+
+void funcB() {
+  UnwindState state;
+  uint64_t regs[K_NB_REGS_UNWIND];
+  size_t stack_size = save_context(regs, stack);
```
Comment (r1viollet) [inline test/savecontext-ut.cc:24]: I thought span construction was explicit. TIL

File: src/savecontext.cc [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    asm (
+    // Retrieve caller save registers
+    // Callee save register are not needed since they could contain anything after function return
+    SAVE_REG(RBX)
+    SAVE_REG(RBP)    
+    SAVE_REG(R12)
+    SAVE_REG(R13)
+    SAVE_REG(R14)
+    SAVE_REG(R15)
+    // Bump the stack by 8 bytes to remove the return address,
```
Comment (r1viollet) [inline src/savecontext.cc:45]: thanks for the comments !

**Review by r1viollet — DISMISSED** [review]
Impressive mechanism ! 👏 👏 👏  Can't wait to try it out.
My review does not cover the asm part.

File: LICENSE-3rdparty.csv (hunk 1 above)
Comment (nsavoire) [reply LICENSE-3rdparty.csv:6]: The other solution would be to update to gcc 11 on ubuntu 16.04.

File: test/savecontext-ut.cc (hunk 2 above)
Comment (nsavoire) [reply test/savecontext-ut.cc:24]: Constructors from C arrays / std::array are not explicit, which is nice !

File: src/savecontext.cc (hunk 3 above)
Comment (nsavoire) [reply src/savecontext.cc:45]: Only works for x86_64 for now, I will do the implementation for ARM later.

**Review by r1viollet — APPROVED** [review]
LGTM

**Pull request merged**